### PR TITLE
Experimental improved jukebox [not ready to be merged]

### PR DESCRIPTION
**Pull request by biconou**
According to #276 this is a WIP experimental new jukebox based on Java Sound API. 

- does not use ffmpeg
- plays gapless
- plays Hi-Resolution audio files in Hi-Resolution mode if the audio device supports it
- supported formats : mp3, flac, ogg
- new UI that allows to position in a song

To use the new Jukebox, create a player and choose the "Java jukebox" technology instead of the "Jukebox" one.

Note that this jukebox player uses a java audio player I started to develop here : https://github.com/biconou/AudioPlayer

### CODE REVIEW

**Comment by biconou** [comment]
@muff1nman @EugeneKay  This PR is now reasonably stable to be tested and (I think) can be added to a 6.3 roadmap

**Comment by muff1nman** [comment]
I don't think its a good idea to add anything else into the /repo folder. We are trying to get rid of it, not add to it. All dependencies should go through maven central.

**Comment by biconou** [comment]
@muff1nman  I agree. I'll find a way to publish my AudioPlayer librery to a central repo for the final release of that PR.

**Comment by biconou** [comment]
@muff1nman 
Because Libresonic version has moved to 6.3 and I removed the AudioPlayer library from the local repo, I think that this PR is ready to be merged. 
The new jukebox player is stable enough even though a lot of improvements and optimisations still have to be done. But the core feature is there and ready to be tested.

File: libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueService.java [hunk 1]
```diff
@@ -612,18 +613,33 @@ public PlayQueueInfo sortByAlbum() throws Exception {
...
     }
 
+    private void updateJukebox(Player player, int offset) throws Exception {
+        if (player.getTechnology() == PlayerTechnology.JUKEBOX) {
+            jukeboxService.updateJukebox(player, offset);
+        } else if (player.getTechnology() == PlayerTechnology.JAVA_JUKEBOX) {
+            jukeboxJavaService.updateJukebox(player, offset);
+        }
+    }
+
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueService.java]: This and other occurrences of this check seem to indicate a violation of dependency inversion. Can we not create an interface for interacting with either technology?

File: libresonic-main/src/main/webapp/WEB-INF/jsp/javaJukeboxPlayerControlBar.jspf [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                <div id="javaJukeboxVolumeSlider" style="width:80px;height:4px"></div>
+            </td>
+        </tr>
+    </table>
+</div>
+
+<script type="text/javascript" src="<c:url value='/script/playQueue/javaJukeboxPlayerControlBar.js'/>"></script>
+<script type="text/javascript">
+    initJavaJukeboxPlayerControlBar();
+</script>
```
Comment (muff1nman) [inline libresonic-main/src/main/webapp/WEB-INF/jsp/javaJukeboxPlayerControlBar.jspf]: jspf? Can we keep the names consistent and use jsp?

File: libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueInfo.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 public class PlayQueueInfo {
 
     private final List<Entry> entries;
+    private final int index;
```
Comment (muff1nman) [inline libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueInfo.java]: Why do we need this new field?

**Comment by muff1nman** [comment]
A rebase will be required so we dont end up with the history of those jar files.

**Comment by biconou** [comment]
@muff1nman Thanks for this review. I'll treat each point.

File: libresonic-main/src/main/webapp/WEB-INF/jsp/javaJukeboxPlayerControlBar.jspf (hunk 2 above)
Comment (biconou) [inline libresonic-main/src/main/webapp/WEB-INF/jsp/javaJukeboxPlayerControlBar.jspf]: .jspf is the standard extension for JSP Fragment

http://stackoverflow.com/questions/2081005/what-is-jspf-file-extension-how-to-compile-it

File: libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueInfo.java (hunk 3 above)
Comment (biconou) [inline libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueInfo.java]: It is needed by the playQueue.jsp view related to the new Jukebox player to know what is the index of the song in queue currently played. It is used by the nowPlayingInfos service callback.

File: libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueService.java (hunk 1 above)
Comment (biconou) [inline libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueService.java]: Agree that it's worth improving.

Comment (biconou) [inline libresonic-main/src/main/java/org/libresonic/player/ajax/PlayQueueService.java]: I didn't want to reorganize too much things here. I think we should but it's not the purpose of this PR.
I think that the updateJukebox method of the PlayQueueService was originally a quick and dirty workaround used to quickly add the jukebox feature in Subsonic. But I think that this jukebox feature was not really well designed and that's why it is weak.
The PlayQueueService should not have an updateJukebox method but instead we should use the play, pause, stop etc ... methods to keep PlayQueueService consistent across all players technologies.
To enforce this idea I also consider that the udateJukebox method is particularly messy as it is the unique entry point to all commands available on a jukebox player.
I've thought of reorganizing all these things but I think it's more worth doing when we change the playQueue UI replacing the PlayQueueService which is currently bound to DWR.

**Comment by biconou** [comment]
@muff1nman I removed any tracks of local repo additions by filtering my git repo with 
`git filter-branch --tree-filter 'rm -rf repo/com/github' HEAD`

**Comment by biconou** [comment]
This PR can be discarded as I created #446

**Pull request closed without merging**